### ISSUE TITLE
^ in regex is reevaluated at each change

### ISSUE THREAD

**Issue by pascalpoitras** [weechat/weechat#1521]
## Bug summary

It looks like ^ is reevaluated at each change in a regex in the regex field of a trigger

## Steps to reproduce

1.  `/trigger add mycomm command mycomm "" "/^(a|b)/c/tg_argv1" "/print -core ${tg_argv1}"`
2.  /mycomm ab

this give _cc_ but it should give _cb_.

in sed

1. `echo 'ab' | sed -E 's/^(a|b)/c/g'`

this give _cb_

suppose you want to erase the first non-space sequence plus the space, you can't do this

1. `/trigger addreplace mycomm command mycomm "" "/^[^ ]+ //tg_argv_eol1" "/print -core ${tg_argv_eol1}"`
2. /mycomm allo ca va

because this will erase all non-space sequence except the last one. So for example _Allo ca va_ will become _va_ but again in sed

`echo 'allo ca va' | sed -E 's/^[^ ]+ //g'`

will correctly gives _ca va_

## Current behavior

Looks like ^ is reevaluated with each change

## Expected behavior

it should work like sed

of course i can fix the last problem with `/^[^ ]+ (.*)/${re:1}/tg_argv_eol1`

---

- WeeChat version: WeeChat 2.9-dev (git: v2.8-186-g196a4e28a)
- OS, distribution and version: Debian 10

**Comment by sim642** [comment]
Indeed, interesting. I was lazy and only added a test for the first example, not thinking there would be anything different for the second one. Now looking at it, there seems to be a somewhat different issue arising on top of the first one.

This is specific to the replacement being empty. If the second trigger is used with a non-empty replacement, then that gets correctly handled. It's only when the replacement moves something completely else to the beginning of the string (by removing the prefix) and that still correctly (at least in terms of the `start_offset`) is the true beginning of the string at that point. Although the issue now is that it wasn't the true beginning of the initial string.

This really pokes a whole in the whole concept of iterating regex match and replace to achieve replace all. I'll have to think about how to keep track of the correspondence between offsets of the initial string and the intermediate string.

**Comment by pascalpoitras** [comment]
it works perfectly now
